### DISCUSSION TITLE
Link to files in a directory (local or online repository)

### DISCUSSION THREAD

**Discussion by anesone** [joonaspaakko/audible-library-extractor#104] [Ideas]
I love ALE as a front-end for my library.  I have backed-up my files to .m4b format (with the ASIN in the filename).

Would it be possible to add a Play option that would look in a specified directory and play (via default system player) the file (.m4b or .mp3) with the same ASIN in the file name?
Being able to specify the app to use to play somewhere in a config would be even better, but gravy.

**Comment by CTinMich** [comment]
Just saw your post...  What did you use to backup your audible to .m4b?

**Reply by joonaspaakko** [reply]
The best options are probably:

- [Libation](https://www.getlibation.com/) (Windows - Free)
- [OpenAudible](https://openaudible.org/) (Mac, Windows, Linux - Paid app)

**Comment by joonaspaakko** [comment]
This is something I've had in my mind for a long time, unfortunately it's not an easy one to implement or even justify. 

**tl;dr** ~~it might be possible using a web player~~ It's possible using a web player in the standalone gallery (much like the current sample player), but would it make sense given Audible's own cloud player can already be opened from ALE? A local file is local, but still?

**Some of my thoughts on it:**

- I believe it should be possible to play a local audio file in the browser.
- Can't open it in an app though. If ALE was a standalone app, then it would be a different matter.
- People seem to hate the cloud player in Audible... Maybe it's a loud minority, but hasn't really inspired confidence in me that it would be worth it to basically make my own cloud player for local files.
- But also it would be way easier to just use Audible's own cloud player, which ALE is capable of opening.
- There's also the limitation that in order load up a local file and play it, ALE would need to know the file's exact path name and extension, which actually I'm not super sure if that's feasible in the extension gallery, but at least in the standalone gallery it should be.
- So realistically speaking all files would likely have to be in a single folder, because web pages have no proper/direct access to the file system. I would also have know the exact naming scheme and be able to generate that same name with the data ALE has access to. If I focused on a single tool like OpenAudible or Libation I could perhaps match the filename, in which case you'd just dump all the files in a single folder and it would just "automatically" find each file when requested, but it would possibly leave any other applications out if their naming scheme is different. And if there's some kinda oddity where a file just has a slightly wrong filename, it just wouldn't play.
- It would be simple if files were named by the ASIN, but would most people bewilling to go through like over anything over a hundred files renaming them to match the asin?
- Getting back to opening files outside of the browser. Like I said, it could be possible if ALE was made into an app. But there are just so many hurdles with that. It could also be possible that if there was a companion audio player app that used deep linking, the gallery could possibly have links that open the app/inform it to play the book in itself, which could search a nested folder structure for the file and play it. So really the player would be, we'll an audiobook player and ALE would be just a separate catalog that you could use to launch it and tell it to play certain files. This is basically what the open in app icon button does in ALE, except it only works in mobile devices and obviously is tied to Audible and only play the books from the Audible app... But the concept is largely the same. Not saying this is any more feasible than turning ALE into an app, but it would perhaps make more sense.

> My idea was that the current sample player could try to first play what it assumes is a local full book and if that doesn't exist, fallback to the sample. This way there wouldn't be a need to add another button that only does something in this very specific case where a file exists.

**Reply by joonaspaakko** [reply]
@anesone, continuing with my addition at the bottom... I realized I could give you a quick example. 

1. [Right now in this url](https://joonaspaakko.github.io/ale-test-new/#/), the first 2 books play "local" test files. Everything else plays the sample.
2. Hover over the cover → click the play button.

> The contents of the link may be overwritten at some point in the distant future.

-----

If you want to play around with it, you can download the test gallery files here [ALE-gallery-local-files-test.zip](https://github.com/joonaspaakko/audible-library-extractor/files/9332213/ALE-gallery-local-files-test.zip)

- You can open the `index.html` in the browser
- If the local file can't be found, it plays the sample, so you can try moving or renaming my test files to see how it works.
- You could also test it with full length files. If you want to test with a book you own (and I have), you can find the ASIN by clicking open the book details and looking at the address bar (`&book=B0923BQ8BW`). Otherwise you can obviously rename any full length book's filename to the ASIN for `Orbs` or whatever.

[<img width="400" alt="ALE-gallery-local-files-test" src="https://user-images.githubusercontent.com/1164476/184511496-28953eb0-0a73-4ea1-b6ec-214d9143abdb.png">](https://user-images.githubusercontent.com/1164476/184511496-28953eb0-0a73-4ea1-b6ec-214d9143abdb.png)


So as I've demonstrated here, doing something like this would be suuuper easy, but there are some things to consider.
- Doesn't have speed control (for the weirdos), but could have if I want to spend any time implementing it. Though I seem to remember it making the audio sound a bit off.
- Doesn't remember where you left off. Again with some extra effort it could be added. Progress could be kept in the browsers memory, but there wouldn't be a way to like move it to another browser, so I can't help but wonder if it would even be worth it.
- Right now it looks for the filename with the `ASIN`, but I suppose the filename format could (probably?) be made to support whatever Libation and OpenAudible outputs. So there could be a setting in the standalone gallery export: `Local audio filename: (x) asin, ( ) OpenAudible, () Libation`
- It seems that adding local files to the "extension gallery" isn't possible. Only the standalone gallery.

But yea, even if I deem all of that too much work to bother doing, as it is right now (in this test case) I could totally leave it like that for the next version. It's barebones as heck, but it works.

**Reply by anesone** [reply]
So, my main reason for suggesting this, is now that the Windows Audible app is defunct, we need a good way to listen to books offline.
I like to listen at work, but don't want to use bandwidth to stream using the web player, thus the need for local files.
I just found Audiobook Suite, which seems like it should do a good job as the back-end part (playing books and keeping your position).
I really like the idea of ALE (standalone gallery) being the front-end.

Libation allows you to name the files using multiple variables, so there is no standard way.  I wasn't sure if it was possible for ALE to search a local directory on a partial file name (the ASIN part), that also includes author, title, etc.
I found this page, but don't know if it relates to your project: https://stackoverflow.com/questions/57782231/how-to-create-a-search-tool-that-displays-files-in-a-folder-as-downloadable-link

**Reply by joonaspaakko** [reply]
Unfortunately no. It could only search a directory if it was a standalone app. Same applies to opening another app.

The extent of what ALE can do is a player on the browser in the standalone gallery. And only if the files are located in the root folder and have a specific name it knows to look for. 

Would my example case be enough, if it could also remember where you left off?

I recommend you download the zip file from my previous comment and add just one full length audiobook there and see how it does.

> I could've also sent you this build of ALE so you could've saved your own standalone gallery that can do it, but the current draft of the gallery is broken as heck so it wouldn't have done you much good. It's better if you just test it with my standalone gallery (in the zip).

**Reply by anesone** [reply]
I think if it can remember listening position, yes, it would be a decent alternative to the defunct desktop app.

**Reply by joonaspaakko** [reply]
Alright. ~~I can't promise that for the next version but I'll keep that in the todo-list.~~ (It's done and will be in then ext version v.0.2.9.) — **What is the exact filename format?** 1 full filename example should be enough.

> Getting back to Audible killing the windows app, did you ever try running the Android app on Windows? I haven't, but I'd assume it would work just the same with downloaded files. Should be possible prior to Win 11 too with [some 3rd party app](https://www.bluestacks.com/apps/entertainment/audible-on-pc.html). As long as Amazon hasn't burned to the ground, that would likely be the best option. Obviously that ignores ALE entirely, but still that's what I'd use if I didn't have wireless headphones that can pipe in PC audio and phone audio through bluetooth simultaneously (for gaming purposes). Although at work I use bluetooth headphones that can connect to 2 devices at the same time and if I want to hear audio from my Mac, I pause phone and it picks up Mac audio and vice versa. At work I don't really have a need to hear PC audio when I'm listening to audiobooks or anything else on my phone... vice versa...

**Reply by joonaspaakko** [reply]
@anesone ^

**Reply by anesone** [reply]
I used the following in Libation to create the filenames:
<first author> - <if series-><series>_<series#> - <-if series><title short> [<id>]

Example 1 (series):
Brandon Sanderson - Stormlight Archive_1 - The Way of Kings [B003ZWFO7E].m4b

Example 2 (non-series)
Fredrik Backman - A Man Called Ove [1797108263].m4b

**Comment by anesone** [comment]
I used Libation.

**Comment by anesone** [comment]
I configured the file names like this:
Brandon Sanderson - Stormlight Archive_1 - The Way of Kings [B003ZWFO7E].m4b

That's not a bad idea on the Android emulator for Windows. I've used Nox on
my boys' laptops, so they can play some of their android games on Windows.
I could always install Audible and Kiwi browser, use your extension to
build the standalone library inside the emulator.
That's my project for this week.

**Reply by joonaspaakko** [reply]
Yea, that's a good idea...
